### PR TITLE
Feat/implement ACT-R

### LINKED ISSUE #50: [Feature request] Add the ACT-R model (see paper)
[Relevant paper](http://act-r.psy.cmu.edu/wordpress/wp-content/uploads/2012/12/409s15516709cog0000_14.pdf).
Main formulas:
image
image
The appendix has example calculations.

It only has 4 parameters: decay intercept (a), decay scale (c), activation threshold (𝜏) and noise (s). Seems to be fairly straightforward to implement. In the paper 𝜏 and s are fixed, but I think we should make them optimizable. I don't see why they should be fixed.
Btw, a reminder: please add the Transformer to the table with p-values.

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
waiting for benchmarking results

### CODE REVIEW

**Comment by Expertium** [comment]
According to the paper, noise and threshold are fixed. I wonder why, I couldn't find a good explanation. To me it seems like they should be optimizable too.
Of course, if you want to do this exactly as the paper suggests, that's understandable. But I think that making the other two parameters optimizable would make more sense, even if it's not according to the paper.
image

**Comment by L-M-Sherlock** [comment]
> According to the paper, noise and threshold are fixed. I wonder why, I couldn't find a good explanation. To me it seems like they should be optimizable too.

They are untrainable. I try to optimize them, but their gradients are `nan`.

You can check it via following script:

```python
import torch
import torch.nn as nn
from torch import Tensor


class ACT_R(nn.Module):
    a = 0.0176786766570677  # decay intercept
    c = 0.216967308403809  # decay scale
    s = 0.254893976981164  # noise
    h = 86400 * 0.025  # time scale
    tau = -0.704205679427144  # threshold
    init_w = [a, c, s, tau]

    def __init__(self, w = init_w):
        super().__init__()
        self.w = nn.Parameter(torch.tensor(w, dtype=torch.float32))

    def forward(self, sp: Tensor):
        """
        :param inputs: shape[seq_len, batch_size, 1]
        """
        m = torch.zeros_like(sp, dtype=torch.float)
        m[0] = -torch.inf
        for i in range(1, len(sp)):
            act = torch.log(
                torch.sum(
                    ((sp[i] - sp[0:i]) * self.h).clamp_min(1)
                    ** (-(self.w[1] * torch.exp(m[0:i]) + self.w[0])),
                    dim=0,
                )
            )
            m[i] = act
        return self.activation(m)

    def activation(self, m):
        return 1 / (1 + torch.exp((self.w[3] - m) / self.w[2]))

    def state_dict(self):
        return list(
            map(
                lambda x: round(float(x), 4),
                dict(self.named_parameters())["w"].data,
            )
        )


model = ACT_R()

t_history_1 = [0, 1, 2, 4, 8]
t_history_2 = [0, 3, 3, 3, 3]

sp = torch.stack([torch.cumsum(torch.tensor(t_history_1), dim=0), torch.cumsum(torch.tensor(t_history_2), dim=0)], dim=1)

print(sp.shape)

p = model(sp)
print(p)
loss_fn = nn.BCELoss(reduction="none")
loss = loss_fn(p, torch.ones_like(p)).sum()
print(loss)
loss.backward()
print(model.w.grad)
```

Result:
```
torch.Size([5, 2])
tensor([[0.0000, 0.0000],
        [0.9029, 0.8960],
        [0.9473, 0.9435],
        [0.9578, 0.9597],
        [0.9609, 0.9679]], grad_fn=<MulBackward0>)
tensor(200.4808, grad_fn=<SumBackward0>)
tensor([13.1608,  1.6838,     nan,     nan])
```

By the way, ACT-R's time complexity is O(n^2). So it will take days to optimize, which is longer than FSRS.

**Comment by L-M-Sherlock** [comment]
@Expertium, I fixed the `nan` bug. Now the ACT-R model has four trainable parameters.

Edit: Here is the initial results:

```
Model: ACT-R
Total number of users: 238
Total number of reviews: 7300000
Weighted average by reviews:
ACT-R LogLoss (mean±std): 0.328±0.180
ACT-R RMSE(bins) (mean±std): 0.043±0.038

Weighted average by log(reviews):
ACT-R LogLoss (mean±std): 0.380±0.181
ACT-R RMSE(bins) (mean±std): 0.065±0.051

Weighted average by users:
ACT-R LogLoss (mean±std): 0.383±0.182
ACT-R RMSE(bins) (mean±std): 0.069±0.052

weights: [0.0046, 0.25, 0.5368, -0.9978]
Model: FSRS-4.5
Total number of users: 238
Total number of reviews: 7300000
Weighted average by reviews:
FSRS-4.5 LogLoss (mean±std): 0.302±0.165
FSRS-4.5 RMSE(bins) (mean±std): 0.044±0.032

Weighted average by log(reviews):
FSRS-4.5 LogLoss (mean±std): 0.347±0.161
FSRS-4.5 RMSE(bins) (mean±std): 0.063±0.044

Weighted average by users:
FSRS-4.5 LogLoss (mean±std): 0.349±0.162
FSRS-4.5 RMSE(bins) (mean±std): 0.066±0.046
```

It's weird that ACT-R has good performance measured by RMSE(bins), but worse at log loss.

**Comment by Expertium** [comment]
> I fixed the nan bug. Now the ACT-R model has four trainable parameters.

Nice! 

> By the way, ACT-R's time complexity is O(n^2). So it will take days to optimize, which is longer than FSRS.

Yeah, I kinda figured that it would be O(n^2) the moment I saw a for loop inside a for loop in the other user's comment.

**Comment by Expertium** [comment]
> ACT-R RMSE(bins) (mean±std): 0.047±0.044

Wait, this is surprisingly good, what the heck? I did not expect it to perform so well. I thought RMSE will be >10%.

**Comment by L-M-Sherlock** [comment]
I think it's ridiculous that a model which doesn't differentiate the correct response and incorrect response has such a good performance. I plan to draw the calibration graph.

**Comment by Expertium** [comment]
Btw, is FSRS's time complexity O(n)?

**Comment by L-M-Sherlock** [comment]
> Btw, is FSRS's time complexity O(n)?

Of course. FSRS is like RNN. Its each state only relies on the last state. And it's state space is fixed in each step.

**Comment by L-M-Sherlock** [comment]
I know why ACT-R is good at RMSE(bins). The distribution of prediction of ACT-R is very narrow:









As I mentioned before, RMSE(bins) is cheatable. And this is a case that ACT-R cheats RMSE(bins).

**Comment by Expertium** [comment]
Oh, so it's a real-life cheating algorithm that puts all predictions into one bin. I thought we would never see one, unless someone wanted to troll us.

**Comment by L-M-Sherlock** [comment]
Yeah. I plan to reorder the models by log loss after this branch is merged. Otherwise, the ACT-R will be placed in the first rank.

**Comment by Expertium** [comment]
Out of curiosity, try keeping noise and threshold fixed, like in the paper, and make graphs again.

**Comment by Expertium** [comment]
If it turns out that it still cheats, **_DO NOT INCLUDE IT IN THE BENCHMARK_**. Just close as "Unplanned".

**Comment by L-M-Sherlock** [comment]
> If it turns out that it still cheats, DO NOT INCLUDE IT IN THE BENCHMARK. Just close as "Unplanned".

Could you explain why?

**Comment by Expertium** [comment]
Because it's a cheating algorithm. It would make proper algorithms look bad and give readers a false sense of accuracy.

**Comment by Expertium** [comment]
Anyway, first we need to see some calibration graphs with noise and threshold fixed.

**Comment by L-M-Sherlock** [comment]
Metrics of ACT-R with two trainable parameters;

```
Model: ACT-R
Total number of users: 145
Total number of reviews: 4259965
Weighted average by reviews:
ACT-R LogLoss (mean±std): 0.372±0.173
ACT-R RMSE(bins) (mean±std): 0.052±0.040

Weighted average by log(reviews):
ACT-R LogLoss (mean±std): 0.407±0.184
ACT-R RMSE(bins) (mean±std): 0.076±0.056

Weighted average by users:
ACT-R LogLoss (mean±std): 0.409±0.187
ACT-R RMSE(bins) (mean±std): 0.080±0.058

weights: [0.0293, 0.3822]
Model: FSRS-4.5
Total number of users: 145
Total number of reviews: 4259965
Weighted average by reviews:
FSRS-4.5 LogLoss (mean±std): 0.339±0.157
FSRS-4.5 RMSE(bins) (mean±std): 0.047±0.031

Weighted average by log(reviews):
FSRS-4.5 LogLoss (mean±std): 0.361±0.157
FSRS-4.5 RMSE(bins) (mean±std): 0.064±0.041

Weighted average by users:
FSRS-4.5 LogLoss (mean±std): 0.362±0.158
FSRS-4.5 RMSE(bins) (mean±std): 0.067±0.043
```

Calibration graph:

[ACT-R.zip](https://github.com/open-spaced-repetition/fsrs-benchmark/files/14261068/ACT-R.zip)








By the way, I will keep the code merged even if we don't include the results in the benchmark because I will do more research about it in the future.

**Comment by Expertium** [comment]
Some graphs look much better than others. By "better" I mean "less like it's cheating".

**Comment by Expertium** [comment]
@L-M-Sherlock change these:
```
            w[0] = w[0].clamp(0, 10)
            w[1] = w[1].clamp(0, 10)
```
to:
```
            w[0] = w[0].clamp(0.001, 1)
            w[1] = w[1].clamp(0.001, 1)
```

I wonder if maybe our clampings are too generous. None of the values in the paper exceed 1.

**Comment by L-M-Sherlock** [comment]
> I wonder if maybe our clampings are too generous. None of the values in the paper exceed 1.

I have checked the optimized parameters. None of them exceeds [0, 1].

What's your final conclusion? Should I continue the benchmarking?

**Comment by Expertium** [comment]
> What's your final conclusion? Should I continue the benchmarking?
> ACT-R.zip

Some of the graphs look fine, but others look more like cheating. Overall, I'd say it's alright, continue. Oh, and keep the sorting order by RMSE.

**Comment by Expertium** [comment]
Also, out of curiosity, can you tell me how long it takes to run FSRS and ACT-R on, for example, 100 collections? Just to give me a rough idea how much slower ACT-R is.

**Comment by L-M-Sherlock** [comment]
> Also, out of curiosity, can you tell me how long it takes to run FSRS and ACT-R on, for example, 100 collections? Just to give me a rough idea how much slower ACT-R is.

FSRS: 1000 collections per hour
ACT-R: 400 collections per hour

**Comment by ElementaryZ** [comment]
> Some of the graphs look fine, but others look more like cheating. Overall, I'd say it's alright, continue. Oh, and keep the sorting order by RMSE.

Can you explain the calibration graphs and why you say the algorithm is cheating? They don't look like any calibration graphs I'm familiar with.

**Comment by Expertium** [comment]
So here's how it's done:
1) Group predictions into bins, for example: [0.81, 0.82, 0.83, 0.84, 0.85]
2) Group their corresponding real review outcomes into bins as well: [1, 0, 1, 1, 1]
3) Calculate the averages of both: 0.83 and 0.80, respectively
4) Take (average_predicted - average_real)^2
5) Do this for all bins
6) Calculate the weighted root-mean-square error, with the number of reviews within each bin as weight

The idea is that within each bin, the average of predicted probabilities of recall and average retention should be as close as possible. If the algorithm predicts that there is an x% chance that you will get this stuff right, you should get it right x% of the time.
The problem is that this is cheatable: a cheating algorithm can just output a constant that is equal to the average retention of that user. Then all predictions will fall into just one bin. And within that bin, the average of predictions and the average of real outcomes (aka retention) will be exactly the same, achieving an RMSE of 0%. But in the absence of cheats, this metric is not only very useful for evaluating algorithms, but is also intuitively interpretable, unlike the log-loss.

As you can see here, that's pretty much exactly what ACT-R is doing. The distribution of its predictions is very narrow; it concentrates all predictions within a very narrow range.
image

**Comment by Expertium** [comment]
Here's an even better example:

**Comment by Expertium** [comment]
@L-M-Sherlock I've looked at the graphs some more and changed my mind. We shouldn't include ACT-R in the benchmark after all, it cheats too much. Feel free to experiment with it, of course, but don't include the results in readme, in the table with other algorithms.

**Comment by ElementaryZ** [comment]
If it cheats with this specific measure why does it invalidate the log-loss? Have you tried using other measures of fit for binary classification?

**Comment by Expertium** [comment]
It (probably) doesn't invalidate the log-loss, but it's still not a good idea to include an algorithm that basically just outputs a constant every time.
> Have you tried using other measures of fit for binary classification?

We've tried [ICI](https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/issues/2). If you can think of anything else, that would be interesting.

**Comment by Expertium** [comment]
@ElementaryZ LMSherlock made some graphs for [DASH](https://github.com/open-spaced-repetition/fsrs-benchmark/issues/51#issuecomment-1942017399), this is what a non-cheating algorithm's calibration graphs look like

**Comment by Expertium** [comment]
Btw, @L-M-Sherlock the graphs are rather cluttered. I would suggest removing weighted least squares and LOWESS, or at least weighted least squares
